Poorly worded error when trying to use a closed FeatureReader

### Subject of the issue
When using a closed feature reader 
```
final FeatureReader<VariantContext> reader = AbstractFeatureReader.getFeatureReader(VCFFile.getAbsolutePath(), null, new VCFCodec(), true);
reader.close()
reader.query("chr1", 100,100)
```
 one gets a very non-informative error message:

```
Exception in thread "main" htsjdk.samtools.util.RuntimeIOException: java.io.IOException: Cannot seek on stream based file
	at htsjdk.tribble.readers.TabixIteratorLineReader.readLine(TabixIteratorLineReader.java:47)
	at htsjdk.tribble.TabixFeatureReader$FeatureIterator.readNextRecord(TabixFeatureReader.java:158)
	at htsjdk.tribble.TabixFeatureReader$FeatureIterator.<init>(TabixFeatureReader.java:146)
	at htsjdk.tribble.TabixFeatureReader.query(TabixFeatureReader.java:121)
	at edu.mit.broad.picard.genotype.fingerprint.v2.ReconstructTriosFromVCFs.getFingerprintsFromVcf(ReconstructTriosFromVCFs.java:131)
	at edu.mit.broad.picard.genotype.fingerprint.v2.ReconstructTriosFromVCFs.getFingerprints(ReconstructTriosFromVCFs.java:193)
	at edu.mit.broad.picard.genotype.fingerprint.v2.ReconstructTrios.doWork(ReconstructTrios.java:144)
	at edu.mit.broad.picard.genotype.fingerprint.v2.ReconstructTriosFromVCFs.doWork(ReconstructTriosFromVCFs.java:101)
	at picard.cmdline.CommandLineProgram.instanceMain(CommandLineProgram.java:205)
	at picard.cmdline.PicardCommandLine.instanceMain(PicardCommandLine.java:94)
	at edu.mit.broad.picard.cmdline.PicardPrivateCommandLine.main(PicardPrivateCommandLine.java:24)
Caused by: java.io.IOException: Cannot seek on stream based file 
	at htsjdk.samtools.util.BlockCompressedInputStream.seek(BlockCompressedInputStream.java:286)
	at htsjdk.tribble.readers.TabixReader$IteratorImpl.next(TabixReader.java:382)
	at htsjdk.tribble.readers.TabixIteratorLineReader.readLine(TabixIteratorLineReader.java:45)
	... 10 more
```

And while I realize that one shouldn't be using a closed reader, it will make debugging a lot easier if the error messages were pertinent to the actual error...

--- a/src/main/java/htsjdk/samtools/util/BlockCompressedInputStream.java
+++ b/src/main/java/htsjdk/samtools/util/BlockCompressedInputStream.java
@@ -57,10 +57,12 @@ public class BlockCompressedInputStream extends InputStream implements LocationA
     public final static String INCORRECT_HEADER_SIZE_MSG = "Incorrect header size for file: ";
     public final static String UNEXPECTED_BLOCK_LENGTH_MSG = "Unexpected compressed block length: ";
     public final static String PREMATURE_END_MSG = "Premature end of file: ";
-    public final static String CANNOT_SEEK_STREAM_MSG = "Cannot seek on stream based file ";
+    public final static String CANNOT_SEEK_STREAM_MSG = "Cannot seek a position for a non-file stream";
+    public final static String CANNOT_SEEK_CLOSED_STREAM_MSG = "Cannot seek a position for a closed stream";
     public final static String INVALID_FILE_PTR_MSG = "Invalid file pointer: ";
 
     private InputStream mStream = null;
+    private boolean mIsClosed = false;
     private SeekableStream mFile = null;
     private byte[] mFileBuffer = null;
     private DecompressedBlock mCurrentBlock = null;
@@ -222,6 +224,9 @@ public class BlockCompressedInputStream extends InputStream implements LocationA
         // Encourage garbage collection
         mFileBuffer = null;
         mCurrentBlock = null;
+
+        // Mark as closed
+        mIsClosed = true;
     }
 
     /**
@@ -344,12 +349,20 @@ public class BlockCompressedInputStream extends InputStream implements LocationA
      * Seek to the given position in the file.  Note that pos is a special virtual file pointer,
      * not an actual byte offset.
      *
-     * @param pos virtual file pointer
+     * @param pos virtual file pointer position
+     * @throws IOException if stream is closed or not a file based stream
      */
     public void seek(final long pos) throws IOException {
+        // Must be before the mFile == null check because mFile == null for closed files and streams
+        if (mIsClosed) {
+            throw new IOException(CANNOT_SEEK_CLOSED_STREAM_MSG);
+        }
+
+        // Cannot seek on streams that are not file based
         if (mFile == null) {
             throw new IOException(CANNOT_SEEK_STREAM_MSG);
         }
+
         // Decode virtual file pointer
         // Upper 48 bits is the byte offset into the compressed stream of a
         // block.

--- a/src/test/java/htsjdk/samtools/util/BlockCompressedOutputStreamTest.java
+++ b/src/test/java/htsjdk/samtools/util/BlockCompressedOutputStreamTest.java
@@ -81,6 +81,7 @@ public class BlockCompressedOutputStreamTest extends HtsjdkTest {
         Assert.assertEquals(bcis2.read(buffer), available, "Should read to end of block");
         Assert.assertTrue(bcis2.endOfBlock(), "Should be at end of block");
         bcis2.close();
+        Assert.assertEquals(bcis2.read(buffer), -1, "Should be end of file");
     }
 
     @DataProvider(name = "seekReadExceptionsData")
@@ -89,24 +90,32 @@ public class BlockCompressedOutputStreamTest extends HtsjdkTest {
         return new Object[][]{
                 {HTSJDK_TRIBBLE_RESOURCES + "vcfexample.vcf.truncated.gz", FileTruncatedException.class,
                         BlockCompressedInputStream.PREMATURE_END_MSG + System.getProperty("user.dir") + "/" +
-                                HTSJDK_TRIBBLE_RESOURCES + "vcfexample.vcf.truncated.gz", true, false, 0},
+                                HTSJDK_TRIBBLE_RESOURCES + "vcfexample.vcf.truncated.gz", true, false, false, 0},
                 {HTSJDK_TRIBBLE_RESOURCES + "vcfexample.vcf.truncated.hdr.gz", IOException.class,
                         BlockCompressedInputStream.INCORRECT_HEADER_SIZE_MSG + System.getProperty("user.dir") + "/" +
-                                HTSJDK_TRIBBLE_RESOURCES + "vcfexample.vcf.truncated.hdr.gz", true, false, 0},
+                                HTSJDK_TRIBBLE_RESOURCES + "vcfexample.vcf.truncated.hdr.gz", true, false, false, 0},
                 {HTSJDK_TRIBBLE_RESOURCES + "vcfexample.vcf.gz", IOException.class,
-                        BlockCompressedInputStream.CANNOT_SEEK_STREAM_MSG, false, true, 0},
+                        BlockCompressedInputStream.CANNOT_SEEK_STREAM_MSG, false, true, false, 0},
+                {HTSJDK_TRIBBLE_RESOURCES + "vcfexample.vcf.gz", IOException.class,
+                        BlockCompressedInputStream.CANNOT_SEEK_CLOSED_STREAM_MSG, false, true, true, 0},
                 {HTSJDK_TRIBBLE_RESOURCES + "vcfexample.vcf.gz", IOException.class,
                         BlockCompressedInputStream.INVALID_FILE_PTR_MSG + 1000 + " for " + System.getProperty("user.dir") + "/" +
-                                HTSJDK_TRIBBLE_RESOURCES + "vcfexample.vcf.gz", true, true, 1000 }
+                                HTSJDK_TRIBBLE_RESOURCES + "vcfexample.vcf.gz", true, true, false, 1000 }
         };
     }
 
     @Test(dataProvider = "seekReadExceptionsData")
-    public void testSeekReadExceptions(final String filePath, final Class c, final String msg, final boolean isFile, final boolean isSeek, final int pos) throws Exception {
+    public void testSeekReadExceptions(final String filePath, final Class c, final String msg, final boolean isFile, final boolean isSeek, final boolean isClosed,
+                                       final int pos) throws Exception {
 
         final BlockCompressedInputStream bcis = isFile ?
                 new BlockCompressedInputStream(new File(filePath)) :
                 new BlockCompressedInputStream(new FileInputStream(filePath));
+
+        if ( isClosed ) {
+            bcis.close();
+        }
+
         boolean haveException = false;
         try {
             if ( isSeek ) {
@@ -212,5 +221,6 @@ public class BlockCompressedOutputStreamTest extends HtsjdkTest {
         }
         bcis.close();
         Assert.assertEquals(deflateCalls[0], 3, "deflate calls");
+        Assert.assertEquals(reader.readLine(), null);
     }
 }

